### PR TITLE
feat: [PL-26991]: Added custom validation framework for settings

### PR DESCRIPTION
**Pull request by tiwaritejas**
## Describe your changes
Support for adding custom validations for individual settings that the teams want to onboard.
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/37339)

### CODE REVIEW

**Comment by tiwaritejas** [comment]
trigger ti3

File: 120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/services/SettingValidator.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package io.harness.ngsettings.services;
+
+import io.harness.ngsettings.dto.SettingDTO;
+
+public interface SettingValidator {
+  void validate(String accountIdentifier, SettingDTO oldSetting, SettingDTO newSetting);
```
Comment (meenakshi-harness) [inline 120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/services/SettingValidator.java:6]: Should we have return type as boolean?

Comment (tiwaritejas) [reply 120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/services/SettingValidator.java:6]: If the validation succeeds, the update operation continues and updated setting response is sent. If validation fails, an error is thrown which is then sent in the error response and shown on the UI. I do not feel the need for a boolean return type here.

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Failed_** 
 


            
 #### 1 Issues 
* Code issue 1 Code issues 

 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/37339)

**Review by nishantsaini20 — CHANGES_REQUESTED** [review]
> Reviewed 4 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37339)* status: Some of the checks are too old. 7 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/services/impl/validators/BaseSettingValidator.java` line 15 at r2](https://reviewable.io/reviews/harness/harness-core/37339#-NCIxyOw1sHcSkcQkgr1:-NCIxyOw1sHcSkcQkgr2:bmld7np) ([raw file](https://github.com/harness/harness-core/blob/9182fab76c97a3b17314b8ffed62c387d85f5aeb/120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/services/impl/validators/BaseSettingValidator.java#L15)):*
> ```Java
> @OwnedBy(PL)
> @Slf4j
> public class BaseSettingValidator implements SettingValidator {
> ```

Rename class to SettingValidatorImpl and mark it as final class.
___
*[`120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/utils/SettingValidationIdentifiers.java` line 3 at r2](https://reviewable.io/reviews/harness/harness-core/37339#-NCIykaI2UCimH87a5He:-NCIykaI2UCimH87a5Hf:b-w3nz2y) ([raw file](https://github.com/harness/harness-core/blob/9182fab76c97a3b17314b8ffed62c387d85f5aeb/120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/utils/SettingValidationIdentifiers.java#L3)):*
> ```Java
> package io.harness.ngsettings.utils;
> 
> public class SettingValidationIdentifiers {}
> ```

Rename class to SettingIdentifiers. Class should be in contracts. Include all the identifiers.

**Review by ankit-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37339)* status: 9 file(s) to review. 2 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/services/impl/validators/SettingValidatorImpl.java` line 16 at r4](https://reviewable.io/reviews/harness/harness-core/37339#-NCJeemI40P_9FhAlquS:-NCJeemI40P_9FhAlquT:b-dwi7dd) ([raw file](https://github.com/harness/harness-core/blob/b2651839d6b834de4de6c57e41605abd0c9ef469/120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/services/impl/validators/SettingValidatorImpl.java#L16)):*
> ```Java
> @Slf4j
> public class SettingValidatorImpl implements SettingValidator {
>   private final Map<String, SettingValidator> settingValidatorMap;
> ```

Whats this map?
___
*[`120-ng-manager/src/main/java/io/harness/ng/NGSettingModule.java` line 34 at r4](https://reviewable.io/reviews/harness/harness-core/37339#-NCJeHWlCNmKTLhMCj-G:-NCJeHWlCNmKTLhMCj-H:b-8hm2b7) ([raw file](https://github.com/harness/harness-core/blob/b2651839d6b834de4de6c57e41605abd0c9ef469/120-ng-manager/src/main/java/io/harness/ng/NGSettingModule.java#L34)):*
> ```Java
>   public void registerSettingValidators() {
>     MapBinder<String, SettingValidator> settingValidatorMapBinder =
>         MapBinder.newMapBinder(binder(), String.class, SettingValidator.class);
> ```

What's this String.class?

**Review by ankit-harness — DISMISSED** [review]
:lgtm:

> Reviewed 1 of 7 files at r1, 1 of 1 files at r2, 2 of 5 files at r3, 2 of 7 files at r5, 5 of 5 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37339#-:-NCOLtyPEGJxiNPMqJvI:bnfp4nl)* status: 6 file(s) to review. 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained.

**Comment by tiwaritejas** [comment]
trigger compile

**Comment by tiwaritejas** [comment]
tigger ti1

**Comment by tiwaritejas** [comment]
trigger ti4

**Review by nishantsaini20 — COMMENTED** [review]
> Reviewed 1 of 1 files at r2, 2 of 5 files at r3, 2 of 7 files at r5, 3 of 5 files at r6, 2 of 2 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37339)* status: 3 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained.

**Review by nishantsaini20 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37339)* status: 3 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by VikasMaddukuriHarness — APPROVED** [review]
> Reviewed 1 of 7 files at r1, 1 of 1 files at r2, 1 of 7 files at r5, 2 of 5 files at r6, 2 of 2 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37339)* status: 1. unresolved discussions.

**Review by tiwaritejas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37339)* status: 3. unresolved discussions. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/NGSettingModule.java` line 34 at r4](https://reviewable.io/reviews/harness/harness-core/37339#-NCJeHWlCNmKTLhMCj-G:-NCf-OaPBSZyptN_F62X:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/b2651839d6b834de4de6c57e41605abd0c9ef469/120-ng-manager/src/main/java/io/harness/ng/NGSettingModule.java#L34)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

What's this String.class?
</blockquote></details>

Done.
___
*[`120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/services/impl/validators/BaseSettingValidator.java` line 15 at r2](https://reviewable.io/reviews/harness/harness-core/37339#-NCIxyOw1sHcSkcQkgr1:-NCJexEuBwoFQPWLWILx:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/9182fab76c97a3b17314b8ffed62c387d85f5aeb/120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/services/impl/validators/BaseSettingValidator.java#L15)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

Rename class to SettingValidatorImpl and mark it as final class.
</blockquote></details>

Done.
___
*[`120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/services/impl/validators/SettingValidatorImpl.java` line 16 at r4](https://reviewable.io/reviews/harness/harness-core/37339#-NCJeemI40P_9FhAlquS:-NCJeyy0A4HEe9nAlEds:b-hoxqmb) ([raw file](https://github.com/harness/harness-core/blob/b2651839d6b834de4de6c57e41605abd0c9ef469/120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/services/impl/validators/SettingValidatorImpl.java#L16)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

Whats this map?
</blockquote></details>

Here, we will maintain a map of settingIdentifier with 
___
*[`120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/utils/SettingValidationIdentifiers.java` line 3 at r2](https://reviewable.io/reviews/harness/harness-core/37339#-NCIykaI2UCimH87a5He:-NCJey9eBkwhDQRb_MzO:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/9182fab76c97a3b17314b8ffed62c387d85f5aeb/120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/utils/SettingValidationIdentifiers.java#L3)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

Rename class to SettingIdentifiers. Class should be in contracts. Include all the identifiers.
</blockquote></details>

Done.

**Review by meenakshi-harness — APPROVED** [review]
> Reviewed 1 of 7 files at r1, 1 of 1 files at r2, 2 of 5 files at r3, 2 of 7 files at r5, 3 of 5 files at r6, 2 of 2 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37339)* status: 2. unresolved discussions.

**Pull request merged**